### PR TITLE
Toggle menuitems

### LINKED ISSUE #8158: add a way to add toggle menu items from php / output/url states
We currently have this js:
'elgg.ui.registerTogglableMenuItems('feature', 'unfeature');'

can't we come up with a way to autodetect those with a generic js function? or register the menu items as toggleable from php?

This could prevent adding js files or have inline js for just toggleable menu items.

Any preference?

### PR DESCRIPTION
**Pull request by jdalsem**
### feat(js): adds param to menuitem to toggle other menu items

fixes #8158

BREAKING CHANGE:
Toggleable menu items are no longer registered with
elgg.ui.registerTogglableMenuItems. Use a 'toggle' attribute when
registering menu items.
### feat(js): added a hook trigger to toggling of menu items

fixes #9697

### CODE REVIEW

**Comment by jdalsem** [comment]
LOL i really like this PR... looking at some of my crappy code for toggling menu items and doing some stuff when a menu item toggles... this helps so much... hope you agree ;)

**Comment by hypeJunction** [comment]
See this: https://github.com/hypeJunction/Elgg-ui_element_states/blob/master/README.md
That's the direction I had in mind. I still don't like having two items in the markup.

File: engine/classes/ElggMenuItem.php [hunk 1]
```diff
@@ -152,6 +156,14 @@ public static function factory($options) {
 			$item->setItemClass($options['item_class']);
 			unset($options['item_class']);
 		}
+		
+		if (isset($options['toggle'])) {
+			// add dependency with toggle AMD
+			if (!isset($options['deps'])) {
+				$options['deps'] = [];
+			}
+			$options['deps'][] = 'js/navigation/menu/elements/item_toggle';
```
Comment (mrclay) [inline engine/classes/ElggMenuItem.php]: Drop the `js/` prefix.

**Comment by mrclay** [comment]
Agree this is better than what we have, but it seems like it doesn't need any integration into the menu system:

```php
ElggMenuItem::factory(array(
	...
	'data-menu-toggle' => 'unfeature',
	'deps' => 'elgg/menu-toggle',
));
```

**Comment by jdalsem** [comment]
> Agree this is better than what we have, but it seems like it doesn't need any integration into the menu system:

why not help people? your proposal will have people wondering why a toggle sometimes works and sometimes not... reason: because they forgot the dependency and another toggle menu item registers it for them... but not in all cases...

**Comment by jdalsem** [comment]
same as the lightbox experience

**Comment by mrclay** [comment]
@hypeJunction can you make a PR for the simplest version of the states model so we can compare?

On this, meeting halfway would be to have core UI code sniff DOM for `[data-menu-toggle]` and auto-require its module. Then at least the menu API doesn't have to learn about this component.

I want to go on record as against helping people :) Well, really 1) sometimes more boilerplate is worth it if it helps devs understand the mechanism, and 2) it feels dirty to start teaching a menu system about client-side behavior modules; I'd rather it stay real dumb.

**Comment by mrclay** [comment]
> core UI code sniff DOM for [data-menu-toggle] and auto-require its module

That doesn't help ajax-delivered markup though :(

**Comment by jdalsem** [comment]
> That doesn't help ajax-delivered markup though :(

thats why i added the hook

**Comment by hypeJunction** [comment]
I will put together something simplistic.

**Comment by jdalsem** [comment]
replaced by #11240

**Pull request closed without merging**